src/misc.rs

```diff
@@ -47,6 +50,190 @@ where
+    let polynomial_iterator = polynomial.iter().enumerate();
+    let mut result = F::zero();
+    let tensor_x = tensor(x);
+    for (i, polynomial_i) in polynomial_iterator {
+        result += tensor_x[i] * polynomial_i;
+    }
+    result
+}
+
+#[test]
+fn test_evaluate_multi_poly() {
+    use crate::ark_std::One;
+    use ark_bls12_381::Fr;
+    let coefs_ones = [Fr::one(); 16];
+    let point_ones = [Fr::one(); 4];
+    let mut result = evaluate_multi_poly(&coefs_ones, &point_ones);
+    assert_eq!(result, Fr::from(16));
+    println!("result: {}", evaluate_multi_poly(&coefs_ones, &point_ones));
+
+    // This evaluation should be 112. I chose these
+    let coefs = [0_u64, 6, 9, 2, 4, 8, 5, 1];
+    let point = [4_u64, 2, 1];
+    // How can I turn this into a function
+    let coefs_fr = coefs.into_iter().map(|x| Fr::from(x)).collect::<Vec<_>>();
+    let point_fr = point.into_iter().map(|x| Fr::from(x)).collect::<Vec<_>>();
+
+    let expected_evaluation = coefs_fr[0]
+        + coefs_fr[1] * point_fr[0]
+        + coefs_fr[2] * point_fr[1]
+        + coefs_fr[3] * point_fr[0] * point_fr[1]
+        + coefs_fr[4] * point_fr[2]
+        + coefs_fr[5] * point_fr[2] * point_fr[0]
+        + coefs_fr[6] * point_fr[2] * point_fr[1]
+        + coefs_fr[7] * point_fr[2] * point_fr[1] * point_fr[0];
+
+    result = evaluate_multi_poly(&coefs_fr, &point_fr);
+    assert_eq!(result, expected_evaluation);
+}
+
+/// divide polynomials
+/// for evaluation point \\(alpha_1, alpha_2, ... alpha_n\\)
+/// get some polynomials \\(q_1, ... q_n\\) such that polynomial = \\(\sum_i{q_i(x)(x_i - \alpha_i)} + f(\alpha)\\)
+/// polynomial is ordered f0, f1, f2, .... f_N
+/// Take the binary expansion of i: b0, b1, ... bn. If b_j is 1, then multiply f_i by x_j.
+///
+pub fn multi_poly_decompose<F: PrimeField>(polynomial: &[F], eval_point: &[F]) -> (Vec<Vec<F>>, F) {
+    let mut results = Vec::new();
+    for _i in 0..eval_point.len() {
+        results.push(vec![F::zero(); polynomial.len()]);
+    }
+    let mut evaluation = F::zero();
+    for (ind, item) in polynomial.iter().enumerate() {
+        // if the i'th bit of j is 1, that means it shouldn't be in the quotient
+        let mut term = *item;
+        let mut ind_shifted = ind;
+        let _counter = 0;
+        for dim in 0..eval_point.len() {
+            if ind_shifted == 0 {
+                break;
+            }
+            if ind_shifted & 1 == 1 {
+                let dim_bit_off = (ind_shifted >> 1) << (1 + dim);
+                results[dim][dim_bit_off] += term;
+                term *= eval_point[dim];
+            }
+            ind_shifted >>= 1;
+        }
+        evaluation += term;
+    }
+
+    // assert_eq!(evaluate_multi_poly(&remainder, evaluation_point), remainder[0]);
+    (results, evaluation)
+}
+
+/// Converts a sparse_polynomial from ark_poly into a vector of coefficients
+pub fn multi_poly_flatten<F: Field, T: Term>(sparse_polynomial: &SparsePolynomial<F, T>) -> Vec<F> {
```

Comment:
why is this here if it's only for multivariate polynomials?

Reply:
Would you prefer that the all the stuff exclusively used for multilinear polynomials goes in the `multikzg` directory?